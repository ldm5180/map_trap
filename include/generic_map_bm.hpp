#ifndef GENERIC_MAP_BM_HPP
#define GENERIC_MAP_BM_HPP

#include "add_to_map.hpp"

#include <boost/hana.hpp>
#include <cstdlib>

static constexpr auto has_reserve =
    boost::hana::is_valid([](auto &&x) -> decltype((void)x.reserve(0)) {});

template <class map_t> static void BM_map_insert(benchmark::State &state) {
  std::srand(42);
  map_t map{};
  if
    constexpr(has_reserve(map)) { map.reserve(state.range(0)); }

  for (auto _ : state) {
    add_to_map(state.range(0), map);
  }
}

template <class map_t> static void BM_map_iterate(benchmark::State &state) {
  std::srand(42);
  map_t map{};
  if
    constexpr(has_reserve(map)) { map.reserve(state.range(0)); }

  for (auto i = 0; i < state.range(0); ++i) {
    add_to_map(state.range(0), map);
  }

  for (auto _ : state) {
    // Nothing to do to sort them. Can simply iterate.
    typename map_t::mapped_type v{};
    for (const auto &kv : map) {
      benchmark::DoNotOptimize(v = kv.second);
    }
  }
}

template <class map_t> static void BM_map_find(benchmark::State &state) {
  std::srand(42);
  map_t map{};
  if
    constexpr(has_reserve(map)) { map.reserve(state.range(0)); }
  std::vector<typename map_t::key_type> keys{};
  keys.reserve(state.range(0));

  for (auto i = 0; i < state.range(0); ++i) {
    add_to_map_and_key_list(state.range(0), map, keys);
  }

  for (auto _ : state) {
    typename map_t::mapped_type v{};
    const auto key = keys[std::rand() % keys.size()];
    benchmark::DoNotOptimize(v = map[key]);
  }
}

#endif // GENERIC_MAP_BM_HPP
