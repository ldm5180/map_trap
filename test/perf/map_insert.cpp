#include "generic_map_bm.hpp"

#include <benchmark/benchmark.h>
#include <map>

BENCHMARK_TEMPLATE(BM_map_insert, std::map<int, std::string>)
    ->RangeMultiplier(2)
    ->Range(8, 8 << 8);
BENCHMARK_MAIN();
