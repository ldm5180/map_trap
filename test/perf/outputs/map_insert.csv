Run on (4 X 2594.09 MHz CPU s)
2017-10-28 17:24:57
name,iterations,real_time,cpu_time,time_unit,bytes_per_second,items_per_second,label,error_occurred,error_message
"BM_map_insert<std::map<int64_t, int64_t>>/8",100000,6972.61,6971.59,ns,,,,,
"BM_map_insert<std::map<int64_t, int64_t>>/16",99809,16469.8,16465.9,ns,,,,,
"BM_map_insert<std::map<int64_t, int64_t>>/32",23371,33258.5,33249.8,ns,,,,,
"BM_map_insert<std::map<int64_t, int64_t>>/64",10000,65245.4,65139.3,ns,,,,,
"BM_map_insert<std::map<int64_t, int64_t>>/128",6685,135669,135644,ns,,,,,
"BM_map_insert<std::map<int64_t, int64_t>>/256",3066,260319,260249,ns,,,,,
"BM_map_insert<std::map<int64_t, int64_t>>/512",1402,517275,517178,ns,,,,,
"BM_map_insert<std::map<int64_t, int64_t>>/1024",950,1.10112e+06,1.10025e+06,ns,,,,,
"BM_map_insert<std::map<int64_t, int64_t>>/2048",387,2.08199e+06,2.07976e+06,ns,,,,,
