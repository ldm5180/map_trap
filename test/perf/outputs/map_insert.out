Run on (4 X 2594.09 MHz CPU s)
2017-10-28 16:31:05
name,iterations,real_time,cpu_time,time_unit,bytes_per_second,items_per_second,label,error_occurred,error_message
"BM_map_insert<std::map<int, std::string>>/8",100000,8071.22,8073.56,ns,,,,,
"BM_map_insert<std::map<int, std::string>>/16",59726,17160.7,17092.3,ns,,,,,
"BM_map_insert<std::map<int, std::string>>/32",29567,34760.4,34775.7,ns,,,,,
"BM_map_insert<std::map<int, std::string>>/64",10000,70687.4,70689.8,ns,,,,,
"BM_map_insert<std::map<int, std::string>>/128",6193,138058,138125,ns,,,,,
"BM_map_insert<std::map<int, std::string>>/256",2928,270613,270635,ns,,,,,
"BM_map_insert<std::map<int, std::string>>/512",1000,560208,560488,ns,,,,,
"BM_map_insert<std::map<int, std::string>>/1024",704,1.39251e+06,1.39268e+06,ns,,,,,
"BM_map_insert<std::map<int, std::string>>/2048",306,2.57311e+06,2.57432e+06,ns,,,,,
